parquet-column/src/main/java/parquet/schema/Types.java

```diff
@@ -555,6 +628,349 @@ protected GroupType build(String name) {
+
+      public MapGroupValueBuilder<P> value(Type.Repetition repetition) {
+        ((MapBuilder)parent).setKeyType(build("key"));
+        return new MapGroupValueBuilder<P>(parent).repetition(repetition);
+      }
+
+      public MapGroupValueBuilder<P> requiredGroupValue() {
+        return value(Type.Repetition.REQUIRED);
+      }
+
+      public MapGroupValueBuilder<P> optionalGroupValue() {
+        return value(Type.Repetition.OPTIONAL);
+      }
+    }
+
+    public static class MapGroupValueBuilder<P> extends BaseGroupBuilder<P,
+        MapGroupValueBuilder<P>> {
+      private final MapBuilder<P> parent;
+
+      public MapGroupValueBuilder(MapBuilder<P> parent) {
+        super(parent.parent);
+        this.parent = parent;
+      }
+
+      public P named(String name) {
+        parent.setValueType(build("value"));
+        return parent.named(name);
+      }
+
+      @Override
+      protected MapGroupValueBuilder<P> self() {
+        return this;
+      }
+    }
+
+    protected void setKeyType(Type keyType) {
+      this.keyType = keyType;
+    }
+
+    protected void setValueType(Type valueType) {
+      this.valueType = valueType;
+    }
+
+    private Type keyType = null;
+    private Type valueType = null;
+    private Class<P> returnType = null;
+
+    public MapBuilder(P parent) {
+      super(parent);
+    }
+
+    private MapBuilder(Class<P> returnType) {
+      super(returnType);
+      this.returnType = returnType;
+    }
+
+    @Override
+    protected MapBuilder<P> self() {
+      return this;
+    }
+
+    public MapKeyBuilder<P> key(PrimitiveTypeName type) {
+      if (parent != null) {
+        return new MapKeyBuilder<P>(this, type);
+      } else {
+        return new MapKeyBuilder<P>(this, returnType, type);
+      }
+    }
+
+    public MapGroupKeyBuilder<P> groupKey() {
+      return new MapGroupKeyBuilder<P>(this);
+    }
+
+    @Override
+    protected Type build(String name) {
+      if (keyType == null) {
+        keyType = STRING_KEY;
+      }
+      return buildGroup(repetition)
+          .as(OriginalType.MAP)
+          .repeatedGroup()
+              .addFields(keyType, valueType)
+              .named("map")
+          .named(name);
+    }
+
+    public P named(String name) {
+      Preconditions.checkState(originalType == null,
+          "Cannot call \"as\" for a MAP type.");
+      return super.named(name);
+    }
+  }
+
+  public static class ListBuilder<P> extends Builder<ListBuilder<P>, P> {
+    private Type elementType = null;
+
+    public ListBuilder(P parent) {
+      super(parent);
+    }
+
+    public ListBuilder(Class<P> returnType) {
+      super(returnType);
+    }
+
+    @Override
+    protected ListBuilder<P> self() {
+      return this;
+    }
+
+    @Override
+    protected Type build(String name) {
+      Preconditions.checkNotNull(elementType, "List element type");
+      return buildGroup(repetition)
+          .as(OriginalType.LIST)
+          .addField(repeatedGroup()
```

Comment:
Should use `repeatedGroup()` directly instead of `addField(repeatedGroup()...)`, I think.